app/src/main/java/org/mozilla/fenix/home/HomeFragmentStore.kt

```diff
@@ -167,7 +169,13 @@ private fun homeFragmentStateReducer(
                 recentHistory = state.recentHistory.filterOut(recentSearchGroup?.searchTerm)
             )
         }
+        is HomeFragmentAction.RemoveRecentTab -> {
+            state.copy(recentTabs = state.recentTabs.filterNot { it == action.recentTab })
```

Comment:
I think we will need to be able to remove recent tabs (and possibly also recent bookmarks) by their `url` or better even `id`, as opposed to relying on object equality. The reason is that the state could be independently updated (and therefore copied) via an observer e.g.: https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentTabsListFeature.kt#L38

This is very likely to happen for tabs as numerous state changes come in asynchronously while the tab is loading.

Definitely an edge case, but could lead to strange bugs where the user removes a recent tab but the tab they tapped on is out of date relative to the store as the UI hasn't updated yet: https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewHolder.kt#L38

So to avoid this ordering problem might be easiest to allow removal by ID / url.

Reply:
This is also a type of bug we had in various places before where we passed the `TabSessionState` around, which wasn't the most up-to-date one anymore, as the object was changed (and therefore copied) in the store. So we switched our use case methods and various interactions over to using the tab.id instead, and look up the up-to-date tab object via the store.

Reply:
Can the `id` ever be equivalent between a `Tab` and a `SearchGroup`?

If this is a common enough comparison, does it make sense to add a method to check equality by id? Something along these lines:

```
sealed class RecentTab
...
    fun hasSameId(other: RecentTab): Boolean =
        this.resolvedId() == other.resolvedId()
    // or
    fun hasSameId(other: RecentTab): Boolean = when {
        this is Tab && other is Tab -> this.state.id == other.state.id
        this is SearchGroup && other is SearchGroup -> this.tabId == other.tabId
        else -> false
    }
                
    
    private fun resolvedId(): String = when(this) {
        is Tab -> state.id
        is SearchGroup -> tabId
    }
}
```

Reply:
Not currently as we won't display a tab in jump back in if it's also in the most recent group. That's so we don't display it twice. 

But if I read this ticket correctly, we also don't plan to support blocking of groups/search terms right now. So I assume there won't be a remove option for search groups is jump back in anyway? If we display a remove option for a groups in jump back in, what would it do? If we just remove the `RecentTab.SearchGroup` it would be immediately be visible again if home is re-opened. We could remove the entire tab group, just keep the tabs. We have `DisbandSearchGroupAction` for that, but not sure that's what we want.

Reply:
So, for individual tabs we can block the URL and remove it from the store using its ID/url. For removing groups from "Jump Back In" we need to clarify what we want to happen. Removing a group from recently visited removes the corresponding metadata and disbands the group. Maybe we want to do the same here, so we could just re-use that functionality and existing action: https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/java/org/mozilla/fenix/home/recentvisits/controller/RecentVisitsController.kt#L109

Reply:
It was determined that adding a remove option for search groups in jump back in is out-of-scope for this ticket. I opened #23785 as a follow-up to add that functionality.

For now, I updated tabs to filter based on their ID, and bookmarks to filter by their URL. Thanks again for the detailed feedback!

Reply:
OK, perfect!